### PR TITLE
Dev/image processing

### PR DESCRIPTION
**Pull request by dsmtE**
## Description

Improvements related to the imageProcessing node

## Features list

- [X] Allow filtering any images in addition to sfm output data
- [X] Add support of fillholes  from OpenImageIO
- [X] Add support of the bilateral filter from OpenCV
- [X] Add support of CLAHE filtering from OpenCV
- [x] Add inputFolders input for batch images filtering

### CODE REVIEW

File: src/software/utils/main_imageProcessing.cpp [hunk 1]
```diff
@@ -183,9 +183,9 @@ int aliceVision_main(int argc, char * argv[])
         return EXIT_FAILURE;
     }
 
-      // Analyze output path
-    const boost::filesystem::path path(sfmOutputDataFilepath);
-    const std::string outputPath = path.parent_path().string();
+    // check if is sfm data file
+    if(fs::path(sfmInputDataFilename).extension().string() == ".abc")
```
Comment (dsmtE) [inline src/software/utils/main_imageProcessing.cpp]: add .sfm extension

File: src/software/utils/main_imageProcessing.cpp [hunk 2]
```diff
@@ -151,31 +208,8 @@ int aliceVision_main(int argc, char * argv[])
...
-
-            const oiio::ImageBuf inBuf(oiio::ImageSpec(w, h, nchannels, oiio::TypeDesc::FLOAT), image.data());
-            oiio::ImageBuf outBuf(oiio::ImageSpec(nw, nh, nchannels, oiio::TypeDesc::FLOAT), rescaled.data());
-
-            oiio::ImageBufAlgo::resize(outBuf, inBuf);
-
-            image.swap(rescaled);
-        }
-        if (exposureCompensation)
+        // if exposureCompensation is needed for sfmData fIles
```
Comment (natowi) [inline src/software/utils/main_imageProcessing.cpp]: `fIles` minor typo

**Comment by fabiencastan** [comment]
Error on windows:
```
C:\projects\alicevision\src\software\utils\main_imageProcessing.cpp(153): error C2562: 'processImage': 'void' function returning a value [C:\projects\alicevision\build\src\software\utils\aliceVision_utils_imageProcessing.vcxproj]
```

File: src/software/utils/main_imageProcessing.cpp (hunk 2 above)
Comment (dsmtE) [reply src/software/utils/main_imageProcessing.cpp]: Fixed in https://github.com/alicevision/AliceVision/pull/806/commits/4670526623ce19e5c00f8cd128f8eaebc9acd934 .

File: src/software/utils/main_imageProcessing.cpp (hunk 1 above)
Comment (dsmtE) [reply src/software/utils/main_imageProcessing.cpp]: Fixed in https://github.com/alicevision/AliceVision/pull/806/commits/4670526623ce19e5c00f8cd128f8eaebc9acd934 .

File: src/aliceVision/dataio/ImageFeed.cpp [hunk 3]
```diff
@@ -224,13 +225,8 @@ ImageFeed::FeederImpl::FeederImpl(const std::string& imagePath, const std::strin
...
       ALICEVISION_LOG_DEBUG("filePattern: " << filePattern);
       std::string regexStr = filePattern;
-      // escape "."
       boost::algorithm::replace_all(regexStr, ".", "\\.");
-      // recognize # as a digit
-      boost::algorithm::replace_all(regexStr, "#", "[0-9]");
-      // recognize @ as a sequence of digits
-      boost::algorithm::replace_all(regexStr, "@", "[0-9]+");
-      re = regexStr;
+      re = filterToRegex(regexStr);
```
Comment (fabiencastan) [inline src/aliceVision/dataio/ImageFeed.cpp:229]: There is a boost::algorithm::replace_all done outside to remove.

File: src/aliceVision/image/io.cpp [hunk 4]
```diff
@@ -90,7 +90,12 @@ bool isSupported(const std::string &ext)
     for(std::string& extension: extensions)
       supportedExtensions.push_back(extension.insert(0, "."));
   }
+  return supportedExtensions;
+}
 
+bool isSupported(const std::string& ext)
+{
+  const std::vector<std::string> supportedExtensions = getSupportedExtensions();
```
Comment (fabiencastan) [inline src/aliceVision/image/io.cpp:98]: We could add static here:
`static const std::vector<std::string> supportedExtensions = getSupportedExtensions();`

File: src/aliceVision/sfmDataIO/viewIO.cpp [hunk 5]
```diff
@@ -203,5 +201,32 @@ std::shared_ptr<camera::IntrinsicBase> getViewIntrinsic(const sfmData::View& vie
   return intrinsic;
 }
 
+const boost::filesystem::path viewPathFromFolders(const sfmData::View& view, const std::vector<std::string>& Folders)
```
Comment (fabiencastan) [inline src/aliceVision/sfmDataIO/viewIO.cpp:204]: folders

File: src/aliceVision/sfmDataIO/viewIO.cpp [hunk 6]
```diff
@@ -203,5 +201,32 @@ std::shared_ptr<camera::IntrinsicBase> getViewIntrinsic(const sfmData::View& vie
...
+{
+    boost::filesystem::path path = "";
+    for(const std::string& folder : Folders)
+    {
+        path = viewPathFromFolder(view, folder);
+        if(path != "")
```
Comment (fabiencastan) [inline src/aliceVision/sfmDataIO/viewIO.cpp:210]:
```
if(!path.empty())
{
    return path;
}
```

File: src/aliceVision/sfmDataIO/viewIO.cpp [hunk 7]
```diff
@@ -203,5 +201,32 @@ std::shared_ptr<camera::IntrinsicBase> getViewIntrinsic(const sfmData::View& vie
...
+            break;
+    }
+
+    return path;
+}
+
+const boost::filesystem::path viewPathFromFolder(const sfmData::View& view, const std::string& folder)
```
Comment (fabiencastan) [inline src/aliceVision/sfmDataIO/viewIO.cpp:217]: no reason for the return value to be const

File: src/software/utils/main_imageProcessing.cpp [hunk 8]
```diff
@@ -120,128 +339,210 @@ int aliceVision_main(int argc, char * argv[])
...
     {
-        ALICEVISION_LOG_ERROR("The input SfMData file '" << sfmInputDataFilename << "' cannot be read.");
+        ALICEVISION_LOG_ERROR("Invalid scale factor, downscale should be in range [0.0001, 1].");
         return EXIT_FAILURE;
     }
 
-    for (auto & viewIt : sfmData.getViews())
+    // Check if sfmInputDataFilename exist and is recognized as sfm data file
+    const std::string inputExt = boost::to_lower_copy(fs::path(inputExpression).extension().string());
+    static const std::array<std::string, 2> SFMSupportedExtensions = {".sfm", ".abc"};
```
Comment (fabiencastan) [inline src/software/utils/main_imageProcessing.cpp:365]: Start variables with lower case: `sfmSupportedExtensions`

File: src/software/utils/main_imageProcessing.cpp [hunk 9]
```diff
@@ -120,128 +339,210 @@ int aliceVision_main(int argc, char * argv[])
...
         }
-        if (sharpenWidth >= 3.f && sharpenContrast > 0.f)
+        else
         {
-            image::Image<image::RGBfColor> filtered(image.Width(), image.Height());
-            const oiio::ImageBuf inBuf(oiio::ImageSpec(image.Width(), image.Height(), nchannels, oiio::TypeDesc::FLOAT), image.data());
-            oiio::ImageBuf outBuf(oiio::ImageSpec(image.Width(), image.Height(), nchannels, oiio::TypeDesc::FLOAT), filtered.data());
-            oiio::ImageBufAlgo::unsharp_mask(outBuf, inBuf, "gaussian", sharpenWidth, sharpenContrast, sharpenThreshold);
+            // If you try to use both a regex-like filter expression and folders as input
+            if(!inputFolders.empty())
```
Comment (fabiencastan) [inline src/software/utils/main_imageProcessing.cpp:485]: Keep the scope around the log (as it's a define).
```
if(!inputFolders.empty())
{
    ALICEVISION_LOG_WARNING...
}
```

File: src/aliceVision/dataio/ImageFeed.cpp (hunk 3 above)
Comment (dsmtE) [reply src/aliceVision/dataio/ImageFeed.cpp:228]: Fixed in https://github.com/alicevision/AliceVision/pull/806/commits/f7584167ca69f68bb430cb92284b5733df155fd7 .

File: src/aliceVision/image/io.cpp (hunk 4 above)
Comment (dsmtE) [reply src/aliceVision/image/io.cpp]: Fixed in https://github.com/alicevision/AliceVision/pull/806/commits/f7584167ca69f68bb430cb92284b5733df155fd7 .

File: src/aliceVision/sfmDataIO/viewIO.cpp (hunk 5 above)
Comment (dsmtE) [reply src/aliceVision/sfmDataIO/viewIO.cpp]: Fixed in https://github.com/alicevision/AliceVision/pull/806/commits/f7584167ca69f68bb430cb92284b5733df155fd7 .

File: src/aliceVision/sfmDataIO/viewIO.cpp (hunk 6 above)
Comment (dsmtE) [reply src/aliceVision/sfmDataIO/viewIO.cpp]: Fixed in https://github.com/alicevision/AliceVision/pull/806/commits/f7584167ca69f68bb430cb92284b5733df155fd7 .

File: src/aliceVision/sfmDataIO/viewIO.cpp (hunk 7 above)
Comment (dsmtE) [reply src/aliceVision/sfmDataIO/viewIO.cpp]: Fixed in https://github.com/alicevision/AliceVision/pull/806/commits/f7584167ca69f68bb430cb92284b5733df155fd7 .

File: src/software/utils/main_imageProcessing.cpp (hunk 8 above)
Comment (dsmtE) [reply src/software/utils/main_imageProcessing.cpp]: Fixed in https://github.com/alicevision/AliceVision/pull/806/commits/f7584167ca69f68bb430cb92284b5733df155fd7 .

File: src/software/utils/main_imageProcessing.cpp (hunk 9 above)
Comment (dsmtE) [reply src/software/utils/main_imageProcessing.cpp:485]: Fixed in https://github.com/alicevision/AliceVision/pull/806/commits/f7584167ca69f68bb430cb92284b5733df155fd7 .

**Comment by fabiencastan** [comment]
"reconstructedViewsOnly" seems to not be taken into account.

**Comment by dsmtE** [comment]
> "reconstructedViewsOnly" seems to not be taken into account. It still detect missing input on non-reconstructed views.

Indeed, the option was not taken into account as well as the new files passed to inputFolders.
Thanks for the review.
fixed in https://github.com/alicevision/AliceVision/pull/806/commits/76817280d274e91e9dc1d439e6e3b21e19bb711c.

**Comment by fabiencastan** [comment]
Sharpen seems to be broken.
Try `sharpenWidth=3` for instance.

**Comment by dsmtE** [comment]
> Sharpen seems to be broken.
> Try `sharpenWidth=3` for instance.

I've just tested it as well and realized the problem.
The problem may come from the switch to RGBA images (used for fillHoles option).

I will check how the OIIO::ImageBufAlgo::unsharp_mask function works in RGBA and otherwise, converting the image to RGB format to apply this filter may be a solution.

The problem could appear on other options too, have you tested the contrast and the median filter?

**Pull request merged**